File: python-algorithm/ds_7_array_inplace/L380-m.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/4 20:05
# @Author  : No Name

class RandomizedSet:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.d = {}
        self.a = []

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.d:
            return False
        self.d[val] = len(self.a)
        self.a.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.d:
            return False
        # 在数组中 把要删除的值和最后一个交换
        self.a[self.d[val]] = self.a[-1]
        self.d[self.a[-1]] = self.d[val]
        self.a.pop(-1)
        self.d.pop(val)
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        idx = random.randint(0, len(self.a) - 1)
        return self.a[idx]

        # Your RandomizedSet object will be instantiated and called as such:
        # obj = RandomizedSet()
        # param_1 = obj.insert(val)
        # param_2 = obj.remove(val)
        # param_3 = obj.getRandom()

File: python-algorithm/ds_7_array_inplace/L380-m.cpp

//380. 常数时间插入、删除和获取随机元素

// 哈希表+动态数组
//动态数组存储元素值
//哈希表存储存储值到索引的映射。

//因为哈希表只能做到查找插入和查找删除是o(1)； 第三个要求随机返回做不到o(1) 所以需要设计一个
//动态数组用于随机返回，此外删除的时候，动态数组可以把最后一个数字和要删除的数字交换位置，然后数组删除最后一个数，并更新哈希

#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class RandomizedSet {
public:
    /** Initialize your data structure here. */
    vector<int> vec;
    unordered_map<int, int> up1;
    RandomizedSet() {}

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(up1.find(val) != up1.end() )
            return false;
        up1.insert(make_pair(val, vec.size()));
        vec.push_back(val);
        return true;
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(up1.find(val) == up1.end())
            return false;
        vec[up1[val]] = vec.back();
        up1[vec.back()] = up1[val]; //把要删除的元素移动到最末尾
        vec.pop_back();
        up1.erase(val);
        return true;

    }

    /** Get a random element from the set. */
    int getRandom() {
        int sz = vec.size();
        if(sz == 0) return 0;
        int index = rand()%sz;
        return vec[index];

    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */